TimeOfFlight/src/TOFEstimators.cc

```diff
@@ -0,0 +1,435 @@
+#include "marlin/VerbosityLevels.h"
+
+#include "marlin/ProcessorEventSeeder.h"
+#include <marlin/AIDAProcessor.h>
+#include <marlin/Global.h>
+
+
+#include <AIDA/IHistogramFactory.h>
+#include <AIDA/ICloud1D.h>
+#include <AIDA/ICloud2D.h>
+#include <AIDA/IHistogram2D.h>
+
+#include <gsl/gsl_rng.h>
+#include <gsl/gsl_randist.h>
+
+//---- ROOT -----
+#include "TH2F.h" 
+
+using namespace lcio ;
+using namespace marlin ;
+using namespace TOFUtils ;
+
+TOFEstimators aTOFEstimators ;
+
+
+
+TOFEstimators::TOFEstimators() : Processor("TOFEstimators") {
+
+  // modify processor description
+  _description = "TOFEstimators compute some estimators for the time of flight from calorimeter hits" ;
+
+  // register steering parameters: name, description, class-variable, default value
+  registerInputCollection( LCIO::RECONSTRUCTEDPARTICLE,
+			   "ReconstructedParticleCollection" , 
+			   "Name of the ReconstructedParticle collection"  ,
+			   _colNamePFO ,
+			   std::string("PandoraPFOs")
+    );
+
+
+    registerProcessorParameter("MaxLayerNumber",
+			     "Use only calorimeter hits up to MaxLayerNumber in TOF estimators",
+			     _maxLayerNum,
+			     int(100) );
+
+
+  registerProcessorParameter("TimeResolution",
+			     "Assumed time resolution per hit in ps",
+			     _resolution,
+			     float(0.) );
+
+}
+
+
+
+void TOFEstimators::init() { 
+
+  streamlog_out(DEBUG) << "   init called  " << std::endl ;
+
+  // usually a good idea to
+  printParameters() ;
+
+  _nRun = 0 ;
+  _nEvt = 0 ;
+
+  // initialize gsl random generator
+  _rng = gsl_rng_alloc(gsl_rng_ranlxs2);
+
+  Global::EVENTSEEDER->registerProcessor(this);
+
+
+  _TOFNames = { "TOFFirstHit",
+		"TOFClosestHits", "TOFClosestHitsError",
+		"TOFCluster",     "TOFClusterError" } ;
+
+}
+
+
+void TOFEstimators::processRunHeader( LCRunHeader* ) {
+
+  _nRun++ ;
+} 
+
+
+
+void TOFEstimators::processEvent( LCEvent * evt ) { 
+
+
+  // use the global Marlin random seed for this processor
+  gsl_rng_set( _rng, Global::EVENTSEEDER->getSeed(this) ) ;   
+
+  streamlog_out(DEBUG ) << "seed set to " << Global::EVENTSEEDER->getSeed(this) << std::endl;
+
+  streamlog_out(DEBUG2) << "   process event: " << evt->getEventNumber() 
+			<< "   in run:  " << evt->getRunNumber() << std::endl ;
+
+
+  // get the PFO collection from the event if it exists
+  LCCollection* colPFO = nullptr ;
+    
+  try{
+    colPFO = evt->getCollection( _colNamePFO ) ;
+  }
+  catch(lcio::Exception){
+    streamlog_out( DEBUG6 ) << " collection " << _colNamePFO
+			    << " not found in event - nothing to do ... " << std::endl ;
+  }
+
+  if( colPFO->getTypeName() != LCIO::RECONSTRUCTEDPARTICLE ) {
+
+    streamlog_out( ERROR ) << " collection " << _colNamePFO
+			   << " not of type LCIO::RECONSTRUCTEDPARTICLE " << std::endl ;
+
+    colPFO = nullptr ;
+  }
+    
+  if( colPFO != nullptr ){
+    
+
+
+    PIDHandler pidh( colPFO );
+    int algoID = pidh.addAlgorithm( name()  , _TOFNames);
```

Comment:
Maybe we should add a processor parameter for this. If it is not provided, we can jump to a fallback with the processor name. What do you think ?

Reply:
Actually, I think this is more elegant as the Marlin framework requires that different processors of the same type have different names. If you had an additional parameter you'd have to change both consistently in the steering file. What would be the advantage ?